src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureHandler.kt

```diff
@@ -0,0 +1,342 @@
+                path.referenceNameElement?.replace(identifier)
+            }
+        }
+        is RsDotExpr -> {
+            if (usage.methodCall?.identifier?.text != config.name) {
+                usage.methodCall?.identifier?.replace(identifier)
+            }
+        }
+    }
+}
+
+private fun changeVisibility(function: RsFunction, config: RsFunctionSignatureConfig) {
+    if (function.vis?.text == config.visibility?.text) return
+
+    function.vis?.delete()
+
+    val vis = config.visibility
+    if (vis != null) {
+        function.addBefore(vis, function.firstChild)
+    }
+}
+
+private fun changeReturnType(factory: RsPsiFactory, function: RsFunction, config: RsFunctionSignatureConfig) {
+    if (function.returnType != config.returnType) {
+        function.retType?.delete()
+        if (config.returnType !is TyUnit) {
+            val ret = factory.createRetType(config.returnType.renderInsertionSafe())
+            function.addAfter(ret, function.valueParameterList)
+        }
+    }
+}
+
+private fun changeParameters(
+    factory: RsPsiFactory,
+    function: RsFunction,
+    config: RsFunctionSignatureConfig,
+    parameterOps: List<ParameterOperation>
+) {
+    val parameters = function.valueParameterList ?: return
+    val parameterList = parameters.valueParameterList.toList()
+    val originalParameters = parameters.copy() as RsValueParameterList
+
+    fun createType(parameter: Parameter): RsTypeReference =
+        factory.tryCreateType(parameter.type.renderInsertionSafe()) ?: factory.createType("()")
+
+    cycle@ for ((index, data) in config.parameters.zip(parameterOps).withIndex()) {
+        val (parameter, op) = data
+        when (op) {
+            is ParameterOperation.Add -> {
+                deleteItem(parameterList, index)
+
+                val typeReference = createType(parameter)
+                val newParameter = factory.createValueParameter(parameter.patText, typeReference, reference = false)
+                val anchor = findAnchorToInsertItem(parameters, index)
+                insertItemWithComma(factory, newParameter, parameters, anchor)
+            }
+            is ParameterOperation.Move -> {
+                deleteItem(parameterList, index)
+
+                val originalParameter = originalParameters.valueParameterList.getOrNull(op.originalIndex)
+                    ?: continue@cycle
+                val anchor = findAnchorToInsertItem(parameters, index)
+                insertItemWithComma(factory, originalParameter, parameters, anchor)
+            }
+        }
+
+        when (op) {
+            is ParameterOperation.Move, is ParameterOperation.Keep -> {
+                val currentParameter = parameters.valueParameterList[index]
+                if (parameter.pat.text != currentParameter.pat?.text) {
+                    if (parameter.pat is RsPatIdent && currentParameter.pat is RsPatIdent) {
+                        renameParameter(function.project, currentParameter, parameter.patText)
```

Comment:
Hmm, I'm not sure how to unify these two. Arguments are changed differently than parameters. For arguments basically I just remove existing arguments or add empty commas. For parameters I need to shuffle them around, rename them/change their type or add new ones at the end.

Reply:
I suggest to refactor `changeParameters` and `changeArguments` methods [like this](https://github.com/intellij-rust/intellij-rust/commit/7da38ed098062fd24cbd8ba6529bca4f4081bacb). That way:
* There will be common method `fixParametersOrder` (so no logic duplication for reordering parameters)
* We will use a bit more functional approach
* Code is shorter (about 100 lines)

Reply:
Thank you. I thought that it would be too complex to unify the logic, but 7 parameters (of `fixParametersOrder`) isn't that bad :D It looks much cleaner now, I just fixed removing all parameters of methods and modified the behaviour of trailing commas - I suppose that if we keep the trailing comma for multiple parameters, we should also keep them if there is only a single parameter?

Also, there is one test that I'm not sure how to make work.
```kotlin
fun `test add multiple parameters`() = doTest("""
    fn foo/*caret*/(a: u32) {}
    fn bar() {
        foo(0);
    }
""", """
    fn foo(a: u32, b: u32, c: u32) {}
    fn bar() {
        foo(0,,);
    }
""") {
    parameters.add(parameter("b", "u32"))
    parameters.add(parameter("c", "u32"))
}
```
The new code creates a different PSI than the PSI parsed from the test. So I'm not sure how to test this (except for changing the behaviour of the refactoring).

Reply:
> I just fixed removing all parameters of methods

👍

> I suppose that if we keep the trailing comma for multiple parameters, we should also keep them if there is only a single parameter?

Looks like both intellij builtin formatter and rustfmt will anyway remove trailing comma if method is single line (and we convert method with one parameter to single line), so I think we should also remove it

> The new code creates a different PSI than the PSI parsed from the test

Yeah, the root cause is that `foo(1, , )` is parsed not as we expected (not as single `RsCallExpr`). We can workaround this by not adding whitespace between commas, so that auto reformatter will add whitespace and fix PSI on the way. That is change `{ factory.createWhitespace(" ") }` to `{ null }` in `changeArguments`